Allow ES6 based tests using jest

ohadlevy opened the pull request:
jest testing support for miq
    
also add missing tests for component registry and enable eslint configuration for react.

ohadlevy commented:
/cc @vojtechszocs

vojtechszocs commented:
> Jest can be used very similarly to jasmine, but allow us to use babel(ES6) and many of its pluging ecosystem (and ofcourse, its faster).

Jest is much more than that 😄 it's converging to become a test platform, beyond a simple test runner.

Jest does stuff like ES6 to ES5 transpilation using existing Babel configuration, which means you don't have to "webpack + run tests", you just "run tests". Simply put, Jest handles the tests and webpack handles the build.

vojtechszocs commented on package.json:
```diff
@@ -73,5 +78,28 @@
+      "__testing__": true
+    },
+    "transform": {
+      "^.+\\.js$": "babel-jest"
```
Custom `transform` isn't necessary, unless you're using extra preprocessors, [see here](https://github.com/facebook/jest/tree/master/packages/babel-jest#setup).

vojtechszocs commented on package.json:
```diff
@@ -73,5 +78,28 @@
+    },
+    "roots": [
+      "app/javascript/packs"
+    ]
```
I'd explicitly specify which files should be picked up for testing, something like:

```
testMatch: ["**/*.test.js"]
```

vojtechszocs commented on package.json:
```diff
@@ -73,5 +78,28 @@
+      "lcov"
+    ],
+    "moduleNameMapper": {
+      "^.+\\.(png|gif|css|scss)$": "identity-obj-proxy"
```
I'd usually mock binary assets, like images & fonts, like this:

```
"\.(png|jpg|gif|svg|eot|ttf|woff|woff2)$": "jest.mock.static-file.js"
```
where the file mentioned above would contain

```
'use strict';
module.exports = 'static-file-stub';
```

so that when someone `import`s a binary asset in JS code, he gets back the `static-file-stub` string.

TL;DR not sure if `identity-obj-proxy` is worth using on binary assets.

ohadlevy commented:
@vojtechszocs in the meanwhile I'm taking a different approach at https://github.com/ohadlevy/miq_experimental_ui_plugin - please let me know what do you think about the defaults in that project.

current blocker I've seen is that travis doesn't know how to run multiple languages in the same project, e.g. you can either have a ruby matrix or a node matrix - not both.

ohadlevy commented:
@vojtechszocs can you please have another look?

ohadlevy commented:
I've left a failing test so we can see it actually failing, once you approve I'll fix the test.
/cc @martinpovolny

ohadlevy commented on .babelrc:
```diff
@@ -1,5 +1,8 @@
 {
-  "presets": [
-    ["env", { "modules": false } ]
```
any idea why we had modules: false to begin with?

himdel replied:
https://github.com/rails/webpacker/blob/master/lib/install/config/.babelrc#L4 is the only reason I think.

(Mostly our webpack config is just webpacker default.)

ohadlevy commented on .travis.yml:
```diff
@@ -15,6 +15,7 @@ env:
   - TEST_SUITE=spec
   - TEST_SUITE=spec:javascript
   - TEST_SUITE=spec:compile
+  - TEST_SUITE=spec:jest
```
I was unable to find a way to have a different language (besides ruby) for travis to run while keeping the ruby one, this means we execute the tests as a rake task under ruby environment, however I believe this is exactly how jasmine tests run today.

It has a few disadvanteges:
1. setup installs the entire environment, when in fact we only need yarn (adds about 10 minutes to the test)
2. we run jest twice with two different rubies.
3. we can't run with different node version.

@skateman - you once mentioned you know of a way to run travis with different language matrix... any idea? :)

skateman replied:
@ohadlevy yes, look [here](https://github.com/skateman/purr/blob/master/.travis.yml) and [here](https://github.com/skateman/purr/blob/master/ci/before_install). Basically travis provides support for a `generic` container which is close to a clean debian/ubuntu base container. Here you can set up your stuff without interfering with the preinstalled environment.

ohadlevy replied:
@skateman thanks - the way I read it:
1. we can do it, but it means we have to revisit/update how we use travis today.
2. not in this PR :)

himdel commented:
Any chance we can keep the test nicely separated from the code as we have now?

Say, `app/javascript` for code and `spec/javascript` for specs?

ohadlevy commented:
> Any chance we can keep the tests nicely separated from the code as we have now?

Sure we can, but i have a few reason why i suggest this way:

1. this code is complied, in theory, we dont need the app/javascript folder content in our build (so less of a reason if you care about pushing test code to "production")
2. when using ES6, we often import, if the files are in different folders, i have to ```import x from '../../../app/javascripts/some/path/to/file'``` vs ```import x from 'file'```.
3. when reviewing PR's, i see logically the test and the code at the same time (vs scroll down to the tests and up to the code over and over again.

> Otherwise, LGTM, but the test failure is relevant ;).


You mean the hello angular test? I left it on purpose so you can see its actually failing - happy to remove if you ack.

himdel commented:
OK, I don't really have a strong opinion about the separation, I consider it useful because when searching, you usually only care about code, not specs and if those are together, it can get annoying to keep seeing specs in search results.

But I can set my search tool to ignore `*.test.js` files and the only negative effect will be that I won't ever notice stuff mentioned only in specs again :).


> You mean the hello angular test? I left it on purpose so you can see its actually failing - happy to remove if you ack.

Aah, then yup, this looks promising :) It actually looks readable in the travis output..

Feel free to remove, I still want to give this a spin locally to see how it works in a browser, but IMO LGTM.

ohadlevy commented:
@himdel I've removed the faling test - if someone wants to see it have a look at https://travis-ci.org/ManageIQ/manageiq-ui-classic/jobs/334595691

I also added a commit which fixes linting errors due to new the configuration.

himdel commented:
Running the whoe test suite:

`yarn test`


Running a specific test:

`yarn run jest app/javascript/packs/hello_angular.test.js`


Running with debugger enabled:

`node --inspect-brk node_modules/.bin/jest --runInBand`

and then there's a green button in chrome's inspector, or a list in `chrome://inspect/#devices`.

(still hoping to find a way to open this automatically)


.. we should add this info somewhere in guides.

ohadlevy commented:
@himdel added task:
```
rake spec:jest:debug
                                                                                            
open your chrome://inspect/#devices on your chrome based browser (see https://facebook.github.io/jest/docs/en/troubleshooting.html for more details)

Debugger listening on ws://127.0.0.1:9229/1b95392b-d88c-449d-bf0b-dfd93a747940
For help see https://nodejs.org/en/docs/inspector
...
```

himdel commented:
Thanks, waiting for travis.. :)
